### PR TITLE
Implement split_array and split_array_mut

### LINKED ISSUE #74674: const-generic array splitting
Creating a new issue about const-generic array splitting methods since it's been discussed on the [array_chunks PR](https://github.com/rust-lang/rust/pull/74373), where it really doesn't belong.

We probably want

```rust
impl [T] {
    fn split_array<const N: usize>(&self) -> (&[T; N], &[T]) { ... }
}

// requires more const generics work to be accepted by rustc AFAIK
impl<T, const N: usize> [T; N] {
    fn split_array<const M: usize>(&self) -> (&[T; M], &[T; {N - M}]) { ... }
}
```

The possibility of a variadic array splitting function was also mentioned (at least that's my interpretation) but since there's not even an active RFC about variadic generics, I think that's a bit out of scope. The only thing one could do once the second `split_array` above becomes possible is add `split_array_2`, `split_array_3` and so on.

### PR DESCRIPTION
**Pull request by jethrogb**
This implements `[T]::split_array::<const N>() -> (&[T; N], &[T])` and `[T; N]::split_array::<const M>() -> (&[T; M], &[T])` and their mutable equivalents. These are another few “missing” array implementations now that const generics are a thing, similar to #74373, #75026, etc. Fixes #74674.

This implements `[T; N]::split_array` returning an array and a slice. Ultimately, this is probably not what we want, we would want the second return value to be an array of length N-M, which will likely be possible with future const generics enhancements. We need to implement the array method now though, to immediately shadow the slice method. This way, when the slice methods get stabilized, calling them on an array will not be automatic through coercion, so we won't have trouble stabilizing the array methods later (cf. into_iter debacle).

An unchecked version of `[T]::split_array` could also be added as in #76014. This would not be needed for `[T; N]::split_array` as that can be compile-time checked. Edit: actually, since split_at_unchecked is internal-only it could be changed to be split_array-only.

### CODE REVIEW

File: library/core/src/array/mod.rs [hunk 1]
```diff
@@ -530,6 +530,76 @@ impl<T, const N: usize> [T; N] {
...
+    ///
+    /// {
+    ///     let (left, right) = v.split_at::<6>();
+    ///     assert_eq!(left, [1, 2, 3, 4, 5, 6]);
+    ///     assert_eq!(right, []);
+    /// }
+    /// ```
+    // Note: this is not intended to be stabilized in its current form. The
+    // second element of the return value should instead be `&[T; {N - M}]`.
+    #[unstable(feature = "array_split_array", reason = "not intended for stabilization", issue = "74674")]
```
Comment (jethrogb) [inline library/core/src/array/mod.rs:569]: * [ ] Need to create tracking issue if this is approved

**Comment by slightlyoutofphase** [comment]
For the array-based version of this, wouldn't it possibly make more sense to have it be entirely value-based so as to make it useful in a different way than the slice version? [Something like this for example.](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=cb5b707b08b993d9c8f6d1871dc3524b)

**Comment by jethrogb** [comment]
That's pretty neat. That brings up a question about naming. For arrays you could have

* split_array
* split_array_ref
* split_array_mut

But now array::split_array and slice::split_array are even more different.

**Comment by slightlyoutofphase** [comment]
> That's pretty neat. That brings up a question about naming. For arrays you could have
> 
> * split_array
> * split_array_ref
> * split_array_mut
> 
> But now array::split_array and slice::split_array are even more different.

Well, like, arrays already have easy direct access to all the slice methods anyways via `as_slice()` and `as_mut_slice()` (and via the classic direct `&arr[..]` and `&mut arr[..]` syntax of course). I'd say overall the best API might ultimately actually be something like just having:

`[T; N]::split_array<const M: usize>(self) -> ([T; M], [T; N - M])`
and
`[T]::split_array_ref<const M: usize>(&self) -> (&[T; M], &[T; N - M])`
and
`[T]::split_array_mut<const M: usize>(&mut self) -> (&mut [T; M], &mut [T; N - M])`

That way all three types of functionality are available where it's appropriate for them to be, and there's also no name confusion.

**Comment by jethrogb** [comment]
You can't have `[T]::split_array_ref<const M: usize>(&self) -> (&[T; M], &[T; N - M])`

**Comment by slightlyoutofphase** [comment]
> You can't have `[T]::split_array_ref<const M: usize>(&self) -> (&[T; M], &[T; N - M])`

Sorry, I got mixed up with my copy-pasting. It should have been:

`[T; N]::split_array<const M: usize>(self) -> ([T; M], [T; N - M])`
and
`[T]::split_array_ref<const M: usize>(&self) -> (&[T; M], &[T])`
and
`[T]::split_array_mut<const M: usize>(&mut self) -> (&mut [T; M], &mut [T])`

So the second two are your existing slice ones, but with the immutable version called `split_array_ref` instead.

**Comment by jethrogb** [comment]
I think we'd still want the exact-remainder versions on array references. That's still possible with the names you suggest, but for slices now there's a weird naming difference between split_at and split_array_ref.

**Comment by slightlyoutofphase** [comment]
> I think we'd still want the exact-remainder versions on array references. That's still possible with the names you suggest, but for slices now there's a weird naming difference between split_at and split_array_ref.

That's fair enough. I personally don't actually care too much what the names end up being TBH. Primarily my point was just that I think the value based version would be useful to also have for arrays (regardless of what it's called).

**Comment by slightlyoutofphase** [comment]
Ok, here's a complete, perfectly working, Miri-passing `const_trait_impl` implementation of all three of the "exact remainder" array methods. There's definitely no technical barrier preventing any of them from being written with the desired signatures right now.

```rust
#![allow(incomplete_features)]
#![feature(
    box_syntax,
    const_evaluatable_checked,
    const_generics,
    const_intrinsic_copy,
    const_maybe_uninit_as_ptr,
    const_maybe_uninit_assume_init,
    const_mut_refs,
    const_panic,
    const_precise_live_drops,
    const_ptr_offset,
    const_raw_ptr_deref,
    // This feature should have a more general name considering what it actually does IMO.
    const_refs_to_cell,
    const_slice_from_raw_parts,
    const_trait_impl,
    slice_ptr_get,
)]

use core::mem::{self, MaybeUninit};
use core::ptr;

#[inline(always)]
const unsafe fn ptr_slice_to_array_mut<'a, T, const N: usize>(slice: *mut [T]) -> &'a mut [T; N] {
    &mut *(slice.as_mut_ptr() as *mut [T; N])
}

#[inline(always)]
const unsafe fn ptr_slice_to_array_ref<'a, T, const N: usize>(slice: *const [T]) -> &'a [T; N] {
    &*(slice.as_ptr() as *const [T; N])
}

// This is just a stand-in for `[T; N]`, to simulate how `N` *must* be initially declared somewhere
// entirely outside of `split_array` for it to work as it does below.
pub trait ArrayHelper<T, const N: usize> {
    fn split_array<const M: usize>(self) -> ([T; M], [T; N - M]);
    fn split_array_ref<const M: usize>(&self) -> (&[T; M], &[T; N - M]);
    fn split_array_mut<const M: usize>(&mut self) -> (&mut [T; M], &mut [T; N - M]);
}

impl<T, const N: usize> const ArrayHelper<T, N> for [T; N] {
    // It seems that `const_evaluatable_checked` itself makes these "just work" without even needing
    // a `where` clause.
    #[inline]
    fn split_array<const M: usize>(self) -> ([T; M], [T; N - M]) {
        assert!(M <= N, "Bounds check failure in `[T; N]::split_array'!");
        let mut left = MaybeUninit::<[T; M]>::uninit();
        let mut right = MaybeUninit::<[T; N - M]>::uninit();
        let self_ptr = self.as_ptr();
        let left_ptr = left.as_mut_ptr() as *mut T;
        let right_ptr = right.as_mut_ptr() as *mut T;
        unsafe {
            self_ptr.copy_to_nonoverlapping(left_ptr, M);
            self_ptr.add(M).copy_to_nonoverlapping(right_ptr, N - M);
            mem::forget(self);
            (left.assume_init(), right.assume_init())
        }
    }

    #[inline]
    fn split_array_ref<const M: usize>(&self) -> (&[T; M], &[T; N - M]) {
        assert!(M <= N, "Bounds check failure in `[T; N]::split_array_ref'!");
        let self_ptr = self.as_ptr();
        unsafe {
            // `ptr::slice_from_raw_parts` is a `const fn`. `slice::from_raw_parts` is not.
            let left = ptr::slice_from_raw_parts(self_ptr, M);
            let right = ptr::slice_from_raw_parts(self_ptr.add(M), N - M);
            (ptr_slice_to_array_ref(left), ptr_slice_to_array_ref(right))
        }
    }

    #[inline]
    fn split_array_mut<const M: usize>(&mut self) -> (&mut [T; M], &mut [T; N - M]) {
        assert!(M <= N, "Bounds check failure in `[T; N]::split_array_mut'!");
        let self_ptr = self.as_mut_ptr();
        unsafe {
            // `ptr::slice_from_raw_parts_mut` is a `const fn`. `slice::from_raw_parts_mut` is not.
            let left = ptr::slice_from_raw_parts_mut(self_ptr, M);
            let right = ptr::slice_from_raw_parts_mut(self_ptr.add(M), N - M);
            (ptr_slice_to_array_mut(left), ptr_slice_to_array_mut(right))
        }
    }
}

const X1: ([usize; 2], [usize; 4]) = [1, 2, 3, 4, 5, 6].split_array::<2>();
const Y1: ([usize; 0], [usize; 6]) = [1, 2, 3, 4, 5, 6].split_array::<0>();
const Z1: ([usize; 6], [usize; 0]) = [1, 2, 3, 4, 5, 6].split_array::<6>();
const X2: (&[usize; 2], &[usize; 4]) = [1, 2, 3, 4, 5, 6].split_array_ref::<2>();
const Y2: (&[usize; 0], &[usize; 6]) = [1, 2, 3, 4, 5, 6].split_array_ref::<0>();
const Z2: (&[usize; 6], &[usize; 0]) = [1, 2, 3, 4, 5, 6].split_array_ref::<6>();

fn main() {
    // Everything is boxed for the sake of making sure it passes Miri, FYI.
    let a = [box 1, box 2, box 3, box 4, box 5, box 6].split_array::<2>();
    println!("{:?}", a.0);
    println!("{:?}\n", a.1);
    let b = [box 1, box 2, box 3, box 4, box 5, box 6].split_array::<0>();
    println!("{:?}", b.0);
    println!("{:?}\n", b.1);
    let c = [box 1, box 2, box 3, box 4, box 5, box 6].split_array::<6>();
    println!("{:?}", c.0);
    println!("{:?}\n", c.1);
    let d = [box 1, box 2, box 3, box 4, box 5, box 6];
    let e = d.split_array_ref::<2>();
    println!("{:?}", e.0);
    println!("{:?}\n", e.1);
    let f = d.split_array_ref::<0>();
    println!("{:?}", f.0);
    println!("{:?}\n", f.1);
    let g = d.split_array_ref::<6>();
    println!("{:?}", g.0);
    println!("{:?}\n", g.1);
    let mut h = [box 1, box 2, box 3, box 4, box 5, box 6];
    let i = h.split_array_mut::<2>();
    for (l, r) in i.0.iter_mut().zip(i.1.iter_mut()) {
        *l.as_mut() += *r.as_ref();
        *r.as_mut() += *l.as_ref();
    }
    println!("{:?}", i.0);
    println!("{:?}\n", i.1);
    let j = h.split_array_mut::<0>();
    for (l, r) in j.0.iter_mut().zip(j.1.iter_mut()) {
        *l.as_mut() += *r.as_ref();
        *r.as_mut() += *l.as_ref();
    }
    println!("{:?}", j.0);
    println!("{:?}\n", j.1);
    let k = h.split_array_mut::<6>();
    for (l, r) in k.0.iter_mut().zip(k.1.iter_mut()) {
        *l.as_mut() += *r.as_ref();
        *r.as_mut() += *l.as_ref();
    }
    println!("{:?}", k.0);
    println!("{:?}\n", k.1);
    // Make sure constant usage works properly also
    println!("{:?}", X1.0);
    println!("{:?}\n", X1.1);
    println!("{:?}", Y1.0);
    println!("{:?}\n", Y1.1);
    println!("{:?}", Z1.0);
    println!("{:?}\n", Z1.1);
    println!("{:?}", X2.0);
    println!("{:?}\n", X2.1);
    println!("{:?}", Y2.0);
    println!("{:?}\n", Y2.1);
    println!("{:?}", Z2.0);
    println!("{:?}\n", Z2.1);
}
```
[Playground link.](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a317e6e3e13968e2a6103a75055a406d)

**Comment by Dylan-DPC** [comment]
@sfackler any updates?

**Comment by JohnTitor** [comment]
Triage: Re-assigning randomly, r? @yaahc

**Comment by JohnCSimon** [comment]
@yaahc  any updates?

File: library/core/src/array/mod.rs [hunk 2]
```diff
@@ -530,6 +530,88 @@ impl<T, const N: usize> [T; N] {
...
+    ///     let (left, right) = v.split_array::<6>();
+    ///     assert_eq!(left, &[1, 2, 3, 4, 5, 6]);
+    ///     assert_eq!(right, []);
+    /// }
+    /// ```
+    // Note: this is not intended to be stabilized in its current form. The
+    // second element of the return value should instead be `&[T; {N - M}]`.
+    #[unstable(
+        feature = "array_split_array",
+        reason = "not intended for stabilization",
```
Comment (jhpratt) [inline library/core/src/array/mod.rs:573]: This should be the actual reason. The method is intended to be stabilized, just with a different return type.

File: library/core/src/array/mod.rs [hunk 3]
```diff
@@ -530,6 +530,88 @@ impl<T, const N: usize> [T; N] {
...
+
+    /// Divides one array into two at an index.
+    ///
+    /// The first will contain all indices from `[0, M)` (excluding
+    /// the index `M` itself) and the second will contain all
+    /// indices from `[M, N)` (excluding the index `N` itself).
+    ///
+    /// # Panics
+    ///
+    /// Panics if `M > N`.
```
Comment (jhpratt) [inline library/core/src/array/mod.rs:542]: It should be possible to add a static assertion until `[T; { N - M }]` is allowed in return types, no?

File: library/core/src/array/mod.rs [hunk 4]
```diff
@@ -530,6 +530,88 @@ impl<T, const N: usize> [T; N] {
...
+    /// {
+    ///     let (left, right) = v.split_array::<6>();
+    ///     assert_eq!(left, &[1, 2, 3, 4, 5, 6]);
+    ///     assert_eq!(right, []);
+    /// }
+    /// ```
+    // Note: this is not intended to be stabilized in its current form. The
+    // second element of the return value should instead be `&[T; {N - M}]`.
+    #[unstable(
+        feature = "array_split_array",
```
Comment (jhpratt) [inline library/core/src/array/mod.rs:572]: Any particular reason to not use a single feature?

**Review by yaahc — CHANGES_REQUESTED** [review]
In addition to agreeing with the comments made so far by @jhpratt I think agree with @slightlyoutofphase that a by value impl for splitting an array in two would almost certainly be useful. We should plan for it now since if we don't add it someone else probably will soon after these methods are made available on nightly, and there seem to be some non-insignificant forward compatibility hazards here to manage.

After digging into the slice split method naming consistency issue you mentioned my first instinct was to call the by value version of the method simply `split`, but that runs into the same coercion issue that you mentioned in the top level comment, since slice already has a split method.

I'm having trouble coming up with any alternatives other than the `_ref` version suggested by @slightlyoutofphase. My current lean is that it's probably fine to have one of the split methods on `slice` include a `_ref` even though similar by reference methods on slice do not. [We have a history of prioritizing forwards compatibility over internal method naming consistency on a given type](https://github.com/rust-lang/rust/pull/82703#issuecomment-827776826).

For now I'm going to suggest updating the PR to add the by value version for array and rename the by reference versions on array and slice to end in `_ref`, since I currently expect that to be the most likely resolution. I'll also cc @rust-lang/libs-api in case anyone can think of any better naming scheme here, but if nobody replies before we've updated it and resolved all the comments I'll happily r+ the `array_split`/`array_split_ref`/`array_split_mut` version.

Edit: I just realized that we probably still cant add `array_split` because of the `[T: {N - M}]` issue so feel free to ignore that part of the above advice if it's impossible and just do the `_ref` rename.

**Comment by jethrogb** [comment]
Updated:

* Created a tracking issue #90091
* Used consistent naming of split_array, split_array_ref, split_array_mut (even for slices)
* Used `N - M` return types for arrays.

It's not clear to me that the `N - M` return types are useful, at least not in the current form of const generics. The compiler doesn't know that `O = N - M`, and as such you can't use the return type in a place where you expect to be able to use `[T; O]`. You can see this in the doc test where I wasn't able to use `[T; O] as PartialEq<[T; N - M]>`.

**Review by BoxyUwU — CHANGES_REQUESTED** [review]

**Comment by BoxyUwU** [comment]
I don't know why the test started failing that's pretty weird, maybe the error is just happening later on now that `core` has the feature enabled...

**Comment by yaahc** [comment]
I think this is looking good, though I would like to add tests before merging. Here are the relevant tests that I can think of:

- #[should_panic] test for when M > N
- M = 0
- N - M = 0
- 0 < M < N

And of course feel free to add any other tests you can think of. I'll r+ after the tests are added. Thanks again for working on this!

**Pull request merged**